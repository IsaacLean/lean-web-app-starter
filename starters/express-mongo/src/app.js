import compression from "compression";
import express from "express";

const app = express();

if (process.env.NODE_ENV === "production") {
  app.use(compression()); // Middleware that compresses most response bodies
}

app.get("/", (req, res) => res.send("Hello World!"));

export default app;
