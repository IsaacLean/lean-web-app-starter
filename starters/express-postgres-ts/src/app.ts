import compression from 'compression'
import express from 'express'
import helmet from 'helmet'

const app = express()

app.use(helmet()) // Middleware that enhances security by setting HTTP response headers

if (process.env.NODE_ENV === 'production') {
    app.use(compression()) // Middleware that compresses most response bodies
}

app.get('/', (req, res) => res.send('Hello World!'))

export default app
