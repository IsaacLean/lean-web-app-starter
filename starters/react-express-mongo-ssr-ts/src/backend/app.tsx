import compression from 'compression'
import express from 'express'
import path from 'path'
import { StrictMode } from 'react'
import { renderToString } from 'react-dom/server'

import {
    PATH_BACKEND_SRC,
    PATH_BUILD_DEV,
    PATH_FRONTEND_BUILD_DEV,
} from '../../PATHS'
import HelloWorld from '../frontend/HelloWorld'

const BUNDLED_BACK_BUILD_PATH = path.resolve(__dirname)
const BUNDLED_FRONT_BUILD_PATH = path.resolve(__dirname, '../frontend')
const BUNDLED_GENERATED_VIEWS_BUILD_PATH = path.resolve(
    __dirname,
    '../generated-views',
)

const app = express()

// Setup EJS templates
app.set('view engine', 'ejs')

// Use the generated views from the frontend build
const viewDirs = []
if (process.env.NODE_ENV === 'test') {
    viewDirs.push(
        `${PATH_BUILD_DEV}/generated-views`,
        `${PATH_BACKEND_SRC}/views`,
    )
} else {
    viewDirs.push(
        BUNDLED_GENERATED_VIEWS_BUILD_PATH,
        `${BUNDLED_BACK_BUILD_PATH}/views`,
    )
}
app.set('views', viewDirs)

// app.use(helmet()) // Middleware that enhances security by setting HTTP response headers

if (process.env.NODE_ENV === 'production') {
    app.use(compression()) // Middleware that compresses most response bodies
}

app.get('/', (req, res, next) => {
    // Server-side render the React app
    let serverSideRendering

    try {
        serverSideRendering = renderToString(
            <StrictMode>
                <HelloWorld />
            </StrictMode>,
        )
    } catch (err) {
        return next(err)
    }

    return res.render('index', {
        title: 'ljas-react-express-mongo-ssr-ts',
        content: serverSideRendering,
    })
})

// Serve the frontend build directory as static files
let frontBuildPath
if (process.env.NODE_ENV === 'test') {
    frontBuildPath = PATH_FRONTEND_BUILD_DEV
} else {
    frontBuildPath = BUNDLED_FRONT_BUILD_PATH
}
app.use('/', express.static(frontBuildPath))

export default app
