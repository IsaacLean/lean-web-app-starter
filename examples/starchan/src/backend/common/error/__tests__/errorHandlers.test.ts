import request from 'supertest'

import app from '../../../app'
import { ssrErrorHandlerTestInjection } from '../ssrErrorHandlerTestInjection'

jest.mock('../ssrErrorHandlerTestInjection')

const ssrErrorHandlerTestInjectionMock = jest.mocked(
    ssrErrorHandlerTestInjection
)

describe('Not found error handlers', () => {
    it('handles a non-API route with no handler using the regular not found error handler', async () => {
        expect.assertions(2)

        const res = await request(app).get('/definitely-not-found')

        expect(res.status).toBe(404)
        expect(res.text).toMatchSnapshot()
    })

    it('handles an API route with no handler using the API not found error handler', async () => {
        expect.assertions(2)

        const res = await request(app).get('/api/v1/definitely-not-found')

        expect(res.status).toBe(404)
        expect(res.body.errors[0]).toBe(
            '/api/v1/definitely-not-found was not found.'
        )
    })
})

describe('API error handler', () => {
    it('returns a 500 error when the fail route is requested', async () => {
        expect.assertions(2)

        const res = await request(app).get('/api/v1/fail')

        expect(res.status).toBe(500)
        expect(res.body.errors[0]).toBe('Internal server error')
    })
})

describe('SSR error handler', () => {
    it('returns a 500 error when the fail route is requested', async () => {
        expect.assertions(2)

        const res = await request(app).get('/fail')

        expect(res.status).toBe(500)
        expect(res.text).toMatchSnapshot()
    })
})

describe('Global error handler', () => {
    beforeEach(() => ssrErrorHandlerTestInjectionMock.mockReset())

    it('returns a 500 error when the fail route is requested', async () => {
        expect.assertions(2)

        ssrErrorHandlerTestInjectionMock.mockImplementation(() => {
            throw new Error(
                'Trigger the global error handler from the SSR error handler'
            )
        })

        const res = await request(app).get('/fail')

        expect(res.status).toBe(500)
        expect(res.text).toMatchSnapshot()
    })
})
