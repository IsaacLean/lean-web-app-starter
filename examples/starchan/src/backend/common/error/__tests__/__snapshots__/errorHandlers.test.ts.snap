// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global error handler returns a 500 error when the fail route is requested 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8" />
        <title>ljas-starchan</title>
        <meta name="viewport" content="width=device-width,initial-scale=1" />
    </head>
    <body>
        <h1>Internal server error</h1>
        
    </body>
</html>
"
`;

exports[`Not found error handlers handles a non-API route with no handler using the regular not found error handler 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8" />
        <title>ljas-starchan</title>
        <meta name="viewport" content="width=device-width,initial-scale=1" />
    <script defer src="/app.js"></script><link href="/app.css" rel="stylesheet"></head>
    <body>
        <div id="root"><h1>Not found</h1><p><i>/definitely-not-found was not found.</i></p><p><a href="/">Return to the home page.</a></p></div>
        <script>
            window.__PRELOADED_STATE__ = {"api":{"queries":{},"mutations":{},"provided":{},"subscriptions":{},"config":{"online":true,"focused":true,"middlewareRegistered":true,"refetchOnFocus":false,"refetchOnReconnect":false,"refetchOnMountOrArgChange":false,"keepUnusedDataFor":60,"reducerPath":"api"}},"appErrors":[{"heading":"Not found","content":"/definitely-not-found was not found."}],"formError":""}
        </script>
    </body>
</html>
"
`;

exports[`SSR error handler returns a 500 error when the fail route is requested 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8" />
        <title>ljas-starchan</title>
        <meta name="viewport" content="width=device-width,initial-scale=1" />
    <script defer src="/app.js"></script><link href="/app.css" rel="stylesheet"></head>
    <body>
        <div id="root"><h1>Internal server error</h1><p><a href="/">Return to the home page.</a></p></div>
        <script>
            window.__PRELOADED_STATE__ = {"api":{"queries":{},"mutations":{},"provided":{},"subscriptions":{},"config":{"online":true,"focused":true,"middlewareRegistered":true,"refetchOnFocus":false,"refetchOnReconnect":false,"refetchOnMountOrArgChange":false,"keepUnusedDataFor":60,"reducerPath":"api"}},"appErrors":[{"heading":"Internal server error"}],"formError":""}
        </script>
    </body>
</html>
"
`;
